Introduce Hyperledger Foundation's security policy

arsulegai opened the pull request:
TODO: need to work on hyperlinks in markdown.

arsulegai committed:
Introduce Hyperledger Foundation's security policy

The security policy that exists currently do not
clearly differentiate the actions that a project
team must take when a vulnerability is reported.
It merely tells the process for reporting a
vulnerability, having a foundation wide common
structure to address the security issues is key
to a healthy community building.

The document here is a proposal put together
by the security process task force setup by the
Hyperledger TOC.

Signed-off-by: hartm <hart.montgomery@gmail.com>
Signed-off-by: S m, Aruna <arun.s.m.cse@gmail.com>

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+(we note that some projects may have different repositories with different
+security policies).  Please edit this as appropriate for your project and
+delete this top section.  For most projects, this will only involve minor
+edits (look for text in red below).  For projects that have security experts,
```
There is no red text in this document. Do we need to add formatting to allow for the red text to show? Or is there another way to represent this text in markdown?

arsulegai replied:
Thanks for catching that. I made the texts `highlighted`. Possible options in markdown are either to `highlight` text, **bold** text or *italicize* text. I am choosing to `highlight` these sections, please suggest.

hartm replied:
It looks like colored text is complicated in markdown, unfortunately.

tkuhrt replied:
Agree. In other templates, I use <mark> tags.

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+delete this top section.  For most projects, this will only involve minor
+edits (look for text in red below).  For projects that have security experts,
+feel free to edit and tailor to your project’s needs, but be sure that you
+know what you’re doing.  We expect this document to be linked from SECURITY.md
```
```suggestion
know what you’re doing.  We expect this document to be linked from `SECURITY.md` in your project's repositories.
```

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+from the OpenSSF,
+https://github.com/ossf/oss-vulnerability-guide/blob/main/maintainer-guide.md.
+
+Each project will have a security team (for more details, see the section on
+people infrastructure below).  The security team will be comprised of
```
```suggestion
Each project will have a security team (for more details, see the
[people infrastructure](#people-infrastructure) below).  The security team will be comprised of
```

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+security and is responsible for responding to and helping to fix security
+vulnerabilities.
+
+The security team (see below for details) for this project will do the
```
```suggestion
The security team for this project will do the
```
Not sure we need to say this again

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+
+The security team (see below for details) for this project will do the
+following for each reported vulnerability:
+1. Acknowledge receipt (see the section on “Report Intakes” for more) of the
```
```suggestion
1. Acknowledge receipt (see [Report Intakes](#report-intakes] for more) of the
```

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+vulnerability, then the reporter should be informed and this process can be
+halted. If the report is still a regular bug (just not a security
+vulnerability), the reporter should be informed (if necessary) of the regular
+process for reporting bugs, available <where your project has regular bug reports>.
```
```suggestion
process for reporting bugs, available \<where your project has regular bug reports\>.
```
Not sure if this is the correct formatting. The portion in the brackets do not show up in the markdown.

arsulegai replied:
I will make them `highlighted`.

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+particular report is more complex, discuss an embargo period with the reporter.
+The embargo period should be negotiated with the reporter and must not be
+longer than 90 days.
+4. Create a patch for the issue (see the “Private Patch Deployment
+Infrastructure” section).
```
```suggestion
4. Create a patch for the issue (see the [Private Patch Deployment
Infrastructure](#private-patch-deployment-infrastructure) section).
```

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+longer than 90 days.
+4. Create a patch for the issue (see the “Private Patch Deployment
+Infrastructure” section).
+5. Request a CVE for the issue (see the “CNA/CVE Reporting” section).
```
```suggestion
5. Request a CVE for the issue (see the [CNA/CVE Reporting](#cnacve-reporting) section).
```

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+Infrastructure” section).
+5. Request a CVE for the issue (see the “CNA/CVE Reporting” section).
+6. Decide the date of public release.
+7. If applicable, Notify members of the embargo list of the upcoming patch
```
```suggestion
7. If applicable, notify members of the embargo list of the upcoming patch
```

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+8. Cut a new (software) release in which the bug is fixed..
+9. Publicly disclose the issue within 48 hours after the release. It is
+recommended that this is done through Github security advisories (see the
+“(Github) Security Advisories” section for more details).
```
```suggestion
[(Github) Security Advisories](#github-security-advisories) section for more details).
```

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+emergency issue arises.  It can be difficult to track down someone with unique
+knowledge (e.g. in a particular area of cryptography) who is capable of
+fixing a problem in a short period of time.
+1. Each maintainer will be a member of the Hyperledger security email list
```
I am assuming this is each maintainer of the security team and not each maintainer. If so, we should rephrase as:
```suggestion
1. Each security team member will be a member of the Hyperledger security email list
```

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+This project has the following recommended ways to submit security
+vulnerabilities.  While the security team members will do their best to
+respond to bugs disclosed in all possible ways, it is encouraged for bug
+finders to report through the following approved channels:
+The Hyperledger Foundation security list:  security@lists.hyperledger.org.
+To report a security issue, please send an email with the name of the project,
```
```suggestion
finders to report through the following approved channels:

- The Hyperledger Foundation security list:  security@lists.hyperledger.org.

To report a security issue, please send an email with the name of the project,
```

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+versions, and if known, mitigations for the issue.  Note that this is a
+mandatory response channel.  We encourage reporters to default to this channel
+for reporting vulnerabilities if there is no compelling reason to use other
+channels. Github security vulnerability reporting (see below for more github
+information):  https://docs.github.com/en/code-security/security-advisories/guidance-on-reporting-and-writing/privately-reporting-a-security-vulnerability.
```
Should the `see below for more github information` be a link to another section in this document? Or is it referring to this link given here?

hartm replied:
I think it's referring to more stuff below in the document, so yes, it should probably be a link.

arsulegai replied:
This part should have been pointers suggesting the possible options to report a security issue. The first one is the security mailing list, the other option is to use github's security reporting framework (the link following the text has more information on that). This section is acting as a template for project maintainers to define what they follow in their case.

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+
+It is recommended but not required that projects use Github as a CNA since the
+reporting process here can be streamlined with the entire Github security
+advisory process (see below).  It is also considered reasonable to have bug
```
Can we provide a link to where we are referring when we say `see below`?

arsulegai replied:
Sure, done.

tkuhrt commented on governing-documents/security.md:
```diff
@@ -0,0 +1,266 @@
+because they can either help fix the problem (perhaps through testing a fix at
+scale) or they need extra time to help prepare their ecosystem to roll out
+fixes quickly.  Approval is not given lightly:  project leadership
+(maintainers) must be convinced that institutions on the list “need to know
```
```suggestion
(maintainers) must be convinced that institutions on the list “need to know"
```

arsulegai committed:
Apply suggestions from Tracy's code review

Co-authored-by: Tracy Kuhrt <tracy.a.kuhrt@accenture.com>
Signed-off-by: Arun S M <aruna.mohan@walmart.com>

arsulegai committed:
Add links in places for `see below`

Signed-off-by: S m, Aruna <arun.s.m.cse@gmail.com>

denyeart commented on governing-documents/security.md:
```diff
@@ -0,0 +1,268 @@
+
+- The Hyperledger Foundation security list:  security@lists.hyperledger.org.
+
+To report a security issue, please send an email with the name of the project,
+a description of the issue, the steps you took to create the issue, affected
+versions, and if known, mitigations for the issue.  Note that this is a
+mandatory response channel.  We encourage reporters to default to this channel
+for reporting vulnerabilities if there is no compelling reason to use other
+channels. Github security vulnerability reporting (see below for more github
+information):  https://docs.github.com/en/code-security/security-advisories/guidance-on-reporting-and-writing/privately-reporting-a-security-vulnerability.
```
OpenSSF advise at https://github.com/ossf/oss-vulnerability-guide/blob/main/maintainer-guide.md:
"If your project is on GitHub, we recommend enabling privately reporting a security vulnerability"

It appears that most if not all Hyperledger projects have private reporting enabled.

With private reporting enabled, any public user can report a security vulnerability directly to the project:
https://docs.github.com/en/code-security/security-advisories/guidance-on-reporting-and-writing/privately-reporting-a-security-vulnerability

I therefore think we need some more clarification in this section. Should users "Report a security vulnerability" via GitHub, or should they use the email? Which method is preferred and why?

I think the deciding factor will be who we want to see the initial reports. Do we want it to be the specific project maintainers? Or do we want it to be a cross-project Hyperledger security team? If the former we should guide towards the GitHub security advisory reporting. If the latter we should guide towards the email. Regardless of the decision here, we should make this document clear about the preferred method. (Note, I am open to either approach).

hartm replied:
I think we need to let the individual projects decide.  Either one would be fine as long as it is clearly specified.  Is there a reason why it should be one or the other?

Besu, for instance, has a very complicated security intake system (because they get bugs reported through a bunch of Ethereum channels).

hartm replied:
I should also note:  the OpenSSF vulnerability disclosure WG did review this document and didn't have a problem with this part of the text.

denyeart replied:
My preference (at least for Fabric) would be to use GitHub vulnerability reporting.
- Keep everything in one place.
- Leverage the existing and well-known GitHub process flow from report through private fork through publish.
- No need to email representatives of every project for an issue with a single project.

A valid reason to use the security email would be so that Hyperledger staff can easily monitor incoming issues and help route/escalate. I thought @hartm might prefer the email for this reason.

All this being said, I'm still fine letting each project decide. I would simply ask that we make the option more clear in this section.

arsulegai committed:
More formatting and links correction

Beautify using markdown format and structure.

Signed-off-by: S m, Aruna <arun.s.m.cse@gmail.com>

arsulegai committed:
Project team to assist in creating embargo list

Signed-off-by: S m, Aruna <arun.s.m.cse@gmail.com>

arsulegai commented:
Comments from the TOC meeting on 3rd August 2023:
1. To define and publish guidelines for embargo list membership. I added a paragraph letting the project team to define an intake process. As TOC we could provide a reference framework. This can be the next `TODO` activity.
2. To suggest project teams to keep the embargo list discussions out of their regular contributor or maintainer calls. This sounds like an instruction for the project team than a policy. I added a pointer in the project incubation exit criteria.
3. I see @denyeart has already brought up the topic of GitHub security reporting vs choosing to send an email.

arsulegai committed:
Idiomatic formatting for consistency

Signed-off-by: S m, Aruna <arun.s.m.cse@gmail.com>

swcurran commented:
Overall, I like the document. However, I find the mix of project guidance and the document being a template for a SECURITY.MD file project difficult. As a maintainer, I think starting the project/repository specific document from this document will be hard to maintain.  I recommend either a second document that is the template for a project, or putting the template at the bottom of the document, clearly delineated from the policy text. The advantage of a separate document is that as the template evolves, it is relatively easy to compare a projects current policy from the Hyperledger Foundation’s recommend template.

I think the template itself should be relatively short, have a pointer to the Hyperledger Foundation policy, and (as is used here) the highlighting to indicate where a project must make choices in their application of the secrity policy options.

swcurran commented on governing-documents/security.md:
```diff
@@ -0,0 +1,274 @@
+`If the project does not use GitHub security advisories, explain how security
+issues are publicly disclosed.`
+
+## Private Patch Deployment Infrastructure
```
Maintaining a private repo for the long term seems like a lot of overhead. Would it be better to document how to create a private repo for when it is needed?  Not sure exactly how to do that, but I’m thinking this could be generic guidance, that a project could dry run and customize for their repo if deemed important to have ready ahead of time.  E.g. create private fork of repo, change X, Y, Z in the pipeline to produce private artifacts.

ryjones replied:
@swcurran this is done automagically by GitHub on a per-patch basis.

arsulegai committed:
Add summary to the policy

Signed-off-by: S m, Aruna <arun.s.m.cse@gmail.com>

denyeart commented:
@arsulegai I can make the edits I proposed above, but I think you first need to update the PR to "Allow edits from maintainers".

arsulegai commented:
> @arsulegai I can make the edits I proposed above, but I think you first need to update the PR to "Allow edits from maintainers".

Done @denyeart. I keep it disabled by default because it also shares the access to secrets.

swcurran commented:
Per the discussion last week, I've created a [draft PR in the Aries ACA-Py repository](https://github.com/swcurran/aries-cloudagent-python/blob/draft-security-policy/SECURITY.md) that I think is an appropriate basis for a "template" project `SECURITY.md` file. Note that in each section the discussion of options has been removed, and only the reasonable default for all projects is given. For Aries, there is no reason not to follow the defaults.  As can be seen, I did leave in a description of the roles and responsibilities of the security team, along with a few other high level "security vulnerability reporting" notes.

To turn this into a template, I would:

- Convert all `Hyperledger Aries` references to `Hyperledger PROJECT`
- Add a Markdown note pointing to the policy document section that describes the alternatives to the default available to projects such that they still adhere to the policy.

For the policy document, I suggest changing the tone (where necessary) to be a policy document not a template, and having for each section of the template a discussion of the alternatives available for projects not wanting to follow the default.

tkuhrt commented:
Reworked in #167

tkuhrt commented on governing-documents/project-incubation-exit.md:
@arsulegai - we should get the changes here into a separate PR.
